### PR TITLE
String interpolator for Show, without macros

### PR DESCRIPTION
**Pull request by dwijnand**
Non-macro alternative to #1489.

Happy to iterate - feedback welcome.

### CODE REVIEW

File: core/src/main/scala/cats/Show.scala [hunk 1]
```diff
@@ -25,6 +25,15 @@ object Show {
     def show(a: A): String = a.toString
   }
 
+  final case class Shown(override val toString: String) extends AnyVal
+  object Shown {
+    implicit def mat[A](x: A)(implicit z: Show[A]): Shown = Shown(z show x)
+  }
+
+  implicit class ShowInterpolator(val _sc: StringContext) extends AnyVal {
+    def show(args: Shown*): String = StringContext(_sc.parts: _*) s (args: _*)
```
Comment (TomasMikula) [inline core/src/main/scala/cats/Show.scala]: `StringContext(_sc.parts: _*)` can be replaced by `_sc`, am I right?

**Comment by TomasMikula** [comment]
I would exercise the principle of least power here and go with this solution over the macro one. It is much more transparent to the casual reader what is going on.

**Comment by peterneyens** [comment]
This looks great :+1:

That implicit conversion to `Shown` is a nice trick.

Would  `syntax.show` be a better place ? So you can get it using `import cats.syntax.show._` (or `cats.implicits._`) instead of `import cats.Show._` ?

**Comment by dwijnand** [comment]
Probably. Or it's own syntax object. Happy to go either way.

**Comment by travisbrown** [comment]
👍 to supporting both `cats.syntax.show._` and `cats.implicits._` if we go this route (which I'm still not sure I prefer over #1489).

**Comment by adelbertc** [comment]
👍

**Comment by dwijnand** [comment]
`cats.syntax.show._`, `cats.syntax.all._` and `cats.implicits._` are all supported now.

**Comment by dwijnand** [comment]
Looking for another maintainer to :+1: this.

**Comment by kailuowang** [comment]
👍

**Pull request merged**